internal/provider/vaultsecrets/resource_vault_secrets_dynamic_secret.go

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package vaultsecrets
+
+import (
+	"context"
+	"fmt"
+	"maps"
+
+	"github.com/hashicorp/hcp-sdk-go/clients/cloud-vault-secrets/preview/2023-11-28/client/secret_service"
+	"github.com/hashicorp/terraform-plugin-framework-validators/objectvalidator"
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-plugin-framework/schema/validator"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-provider-hcp/internal/clients"
+	"github.com/hashicorp/terraform-provider-hcp/internal/provider/modifiers"
+)
+
+// dynamicSecret encapsulates the HVS provider-specific logic so the Terraform resource can focus on the Terraform logic
+type dynamicSecret interface {
+	readFunc(ctx context.Context, client secret_service.ClientService, secret *DynamicSecret) (any, error)
+	createFunc(ctx context.Context, client secret_service.ClientService, secret *DynamicSecret) (any, error)
+	updateFunc(ctx context.Context, client secret_service.ClientService, secret *DynamicSecret) (any, error)
+	// deleteFunc not necessary, all secrets use the same delete request
+	fromModel(s *DynamicSecret, model any) diag.Diagnostics
+}
+
+// dynamicSecretsImpl is a map of all the concrete dynamic secrets implementations by provider
+// so the Terraform resource can look up the correct implementation based on the resource secret_provider field
+var dynamicSecretsImpl = map[Provider]dynamicSecret{
+	ProviderAWS: &awsDynamicSecret{},
+	ProviderGCP: &gcpDynamicSecret{},
+}
```

Comment:
This map of interface pattern has been used successfully many times in Vault ENT & HVS. It eliminates the proliferation of switch cases so adding new variants does not require modifications in several disjointed locations. You only need to implement the interface and add it to the collection.